2022-11-27.rb

```diff
@@ -0,0 +1,61 @@
+require File.expand_path(File.dirname(__FILE__) + '/assert')
+require 'pry'
+require 'json'
+
+# Level: Medium
+# This problem was asked by Google.
+
+# Given the root to a binary tree, implement serialize(root), which serializes the tree into a string,
+# and deserialize(s), which deserializes the string back into the tree.
+# 
+# For example, given the following Node class:
+# 
+# class Node:
+#     def __init__(self, val, left=None, right=None):
+#         self.val = val
+#         self.left = left
+#         self.right = right
+#
+# The following test should pass:
+# 
+# node = Node('root', Node('left', Node('left.left')), Node('right'))
+# assert deserialize(serialize(node)).left.left.val == 'left.left'
+
+# my Ruby implementation of the Node class given in the instructions in Python
+class Node
+  attr_reader :val, :left, :right
+
+  def initialize(val, left=nil, right=nil)
+    @val = val
+    @left = left
+    @right = right
+  end
+end
+
+######### SOLUTIONS #########
+
+# I'm serializing to JSON, since that's a binary tree serialized format I know, and I can use the JSON parser for the deserialization
+# I'm of course using recursivity, and for such tasks I'm not sure there's any other/better way
+def serialize(root)
+  "{\"val\": \"#{root.val}\", \"left\": #{ serialize_or_null(root.left) }, \"right\": #{ serialize_or_null(root.right) }}"
+end
+
+def serialize_or_null(root)
+  root ? serialize(root) : 'null'
+end
+
+# I'm using the JSON parser to get an object, which makes the deserialization much easier (no need to parse strings)
+# Using the language's built in tools isn't considered cheating here, right?
```

Comment:
I think it is, especially since you have JSON in JSON in JSON etc.

And given that you don't escape double-quotes in your `serialize`, if you do a deeper tree I think you'll notice it's broken.

Reply:
Thanks @byroot. I'll try to implement a solution without using JSON then.
I'll need to think of a nice way to represent a binary tree in string format, that could be easily parsed manually.
I'll also add some tests with deeper nesting.